# 🤖 Assignment 2: Deep Learning vs. Traditional Machine Learning

In this notebook, we explore the differences between traditional Machine Learning (ML) and Deep Learning (DL), specifically in the context of image recognition using the MNIST handwritten digit dataset.

---

### **Objectives**
- Understand when to use DL over ML
- Implement both ML and DL approaches
- Compare performance metrics and explain why DL excels for this task


## 📊 Traditional ML vs Deep Learning

| Aspect                    | Traditional ML                      | Deep Learning                            |
|--------------------------|-------------------------------------|------------------------------------------|
| **Feature Engineering**   | Manual, requires domain expertise   | Automatic (learns features from data)    |
| **Data Requirements**     | Works with small datasets           | Requires large datasets                  |
| **Interpretability**      | Easier to interpret                 | Harder to interpret (black box)          |
| **Computation**           | Light to moderate                   | High, requires GPUs                      |
| **Flexibility (e.g., images)** | Limited for unstructured data    | Very strong for unstructured data (e.g., images, audio) |


## 📥 Load and Prepare MNIST Data

In [1]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical
import numpy as np

In [2]:
# Load MNIST data
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalize data
X_train = X_train / 255.0
X_test = X_test / 255.0

# Flatten for ML
X_train_flat = X_train.reshape(-1, 28*28)
X_test_flat = X_test.reshape(-1, 28*28)

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


## 🧠 Traditional ML: Logistic Regression

In [3]:
# Use only a subset for faster training (optional)
ml_model = LogisticRegression(max_iter=1000)
ml_model.fit(X_train_flat, y_train)

ml_preds = ml_model.predict(X_test_flat)
ml_accuracy = accuracy_score(y_test, ml_preds)
ml_accuracy

0.9259

## 🧠 Deep Learning: Neural Network (Keras)

In [4]:
# Convert labels to categorical for DL
y_train_cat = to_categorical(y_train, 10)
y_test_cat = to_categorical(y_test, 10)

# Define simple neural network
dl_model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile and train
dl_model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
dl_model.fit(X_train, y_train_cat, epochs=5, batch_size=32, verbose=1)

# Evaluate
dl_loss, dl_accuracy = dl_model.evaluate(X_test, y_test_cat)
dl_accuracy

/usr/local/lib/python3.11/dist-packages/keras/src/layers/reshaping/flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 8s 4ms/step - accuracy: 0.8781 - loss: 0.4312
Epoch 2/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 9s 3ms/step - accuracy: 0.9640 - loss: 0.1225
Epoch 3/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 3ms/step - accuracy: 0.9763 - loss: 0.0773
Epoch 4/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 7s 4ms/step - accuracy: 0.9820 - loss: 0.0591
Epoch 5/5
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 10s 4ms/step - accuracy: 0.9864 - loss: 0.0439
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9702 - loss: 0.0955


0.975600004196167

## 📈 Results & Analysis

- **Traditional ML Accuracy**: 92.6 %
- **Deep Learning Accuracy**:  97.6 %

---

### 💬 Why Deep Learning Performs Better

- The DL model automatically **extracts pixel-level patterns** without needing manual feature design.
- It can detect **curves, edges, and textures** important for digit recognition.
- Traditional ML (like Logistic Regression) treats each pixel as a flat input and doesn’t learn spatial hierarchies.
- DL also benefits from GPU acceleration and deeper architectures as data size increases.

---

### ✅ Conclusion

For image-based problems like MNIST, **deep learning outperforms traditional ML** due to its strength in feature extraction and hierarchical learning. However, for structured/tabular data, ML may still be more efficient and interpretable.
